### DISCUSSION TITLE
mongodb

### DISCUSSION THREAD

**Discussion by mikenomitch** [cloudflare/workerd#2721] [NPM Package Requests]
The `mongodb` does not work even with `node_compat_v2` on.

It can be imported, but it does not properly connect to a DB. The package relies on `net.Socket` and `tls.TLSSocket`.

See Issue for more context: https://github.com/cloudflare/workers-sdk/issues/6684
See Blog for more context: https://alexbevi.com/blog/2024/09/11/why-cloudflare-workers-dont-work-with-mongodb/

**Comment by vicb** [comment]
@xofyarg reported an [issue on the unenv repo](https://github.com/unjs/unenv/issues/314) which @pi0 fixed.

It should be integrated in unenv with https://github.com/cloudflare/workers-sdk/pull/6920

**Comment by tobsn** [comment]
/cc @ricky11 @anonrig

it's been forever now... I can't remember when I encountered this the first time but it must be at least 2 years ago.
as new mongodb atlas instances now dont even list the deprecated data api menu tab and the existing instances will only be available for one more year, it's about hot damn time this gets fixed... ;)

thanks!

**Reply by anonrig** [reply]
Working on this - https://github.com/cloudflare/workerd/pull/3315

**Reply by tobsn** [reply]
> Working on this - #3315

thank you! this way we can use cloudflare AI on cloudflare workers and still use our mongodb atlas instance... it will save us a massive amount of time and energy. I assume that will be the same with a lot of other people once they realize the September atlas data API deprecation deadline is approaching ... anyway, thanks again! immensely appreciated!

**Comment by faradaytrs** [comment]
I saw the thread on the forum which says that the only problem is that workers don't support TLS

**Comment by Yusaney** [comment]
I've started a new Mongodb instance, and I can't use it with my Cloudflare workers; I want to use a Cloudflare environment with queues, kv, etc., but I can't use it if workers don't actually work...

**Comment by tobsn** [comment]
/cc @ricky11 @anonrig

December 6th

**Reply by tobsn** [reply]
@irvinebroque

**Reply by ricky11** [reply]
i've had to move most of my workers that require mongodb to a node function (vercel/aws) instead, not a great dx, but its the only way this will work and its unlikely the nativer drive will work on cf workers anytime soon or ever.

I did see this blog on TLS support, but i guess : https://blog.cloudflare.com/mtls-workers/ @alexbevi can chime in.

**Reply by MasFana** [reply]
is there any news about connecting MongoDB with Workers?

**Reply by anonrig** [reply]
It's on my agenda. I'm going to be working on this starting from January.

**Reply by ricky11** [reply]
Amazing, tls support for workers whereby native mongodb drivers would work! that would be such a blessing!

**Reply by alexbevi** [reply]
@anonrig this is great news!

**Reply by Imtiajrex** [reply]
any movement on this @anonrig

**Reply by anonrig** [reply]
I've opened an in progress PR - https://github.com/cloudflare/workerd/pull/3315

**Reply by anonrig** [reply]
TLSSocket and connect() has landed into main. It will be available to the general public in the next couple of weeks.

[the remark this answers is not in the document]
**Reply by anonrig** [reply]
Working on this - https://github.com/cloudflare/workerd/pull/3315

**Comment by DaryeDev** [comment]
How can one use mongodb on Workers, now that node:net is merged?

**Reply by vicb** [reply]
It has been merged in `workerd` but it will still take a few days before the prod infra is updated and then `wrangler` is updated. Once this happens (probably towards the end of next week or the week after), you should be able to just use the `mongodb` package.

**Reply by DaryeDev** [reply]
Great to know, thanks!

**Reply by vicb** [reply]
FYi, this code works on bleeding edge (available in 5-10 days)

![image](https://github.com/user-attachments/assets/67a7ba40-497d-4d5c-a39e-e0111b58e0fe)

**Reply by DaryeDev** [reply]
Perfect, should be just updating the worker library, then. Thanks!

**Reply by raininglemons** [reply]
Is there any where public these kind of releases can be tracked?

**Reply by vicb** [reply]
> Is there any where public these kind of releases can be tracked?

We'll give an update here when the changes are deployed.
The current status is that all changes have been merged,
it will take a few days before this reaches wrangler and prod.

**Comment by vicb** [comment]
As promised, here is an update:

workerd now implements `node:timers` and `node:net` (thanks @anonrig).

wrangler 3.106 which was just released contains the latest version of `workerd`.

Meaning that you should be able to use `mongodb` from your node apps!

Please test and report your experiences!

🙏

edit:

You need to use the `nodejs_compat` compatibility flag with a compatibility date set after `2024-09-23`.

I have tested with:

```toml
# wrangler.toml
compatibility_date = "2024-12-05"
compatibility_flags = ["nodejs_compat"]
```

**Reply by JetLua** [reply]
If I use the SRV protocol: ```mongodb+srv://```, an error occurs: `socket.once is not a function`.

**Reply by gerhardcit** [reply]
A real world test will be a signup with Atlas MongoDB and run this on a deployed worker. 
It feels like `mongodb://localhost`  running it locally somehow does not represent the worker space as well.

**Reply by vicb** [reply]
> If I use the SRV protocol: `mongodb+srv://`, an error occurs: `socket.once is not a function`.

Would you have a minimal repro by any chance?

Thanks!

**Reply by raininglemons** [reply]
If @JetLua can't provide a minimal reproducible example, I can share one later and a valid mongodb uri to a test atlas db instance.

FYI, tested with an older connection string, i.e. mongodb:// and it still gives the same issue. So don't think its related to mongodb+srv specifically

**Reply by JetLua** [reply]
@vicb 

```ts
const client = new MongoClient('mongodb+srv://test:juKMkH0RKoTWihf1@yew.oqbm7gt.mongodb.net/test')

const db = client.db()

console.log(await db.collection('test').countDocuments()) // output: 1
```

**Reply by vicb** [reply]
Thanks for the repro.

Took a quick look: the failing `socket` is an instance of `TLSSocket` which is implemented in `unenv`.

**Reply by anonrig** [reply]
@JetLua @vicb I'll be working on adding TLSSocket.

**Reply by alexbevi** [reply]
@anonrig looking forward to TLSSocket support!

**Reply by DylanBulmer** [reply]
@vicb Thank you for the update! Here's the summary of my findings:

If I use the `mongodb://` connection string protocol, I have the error most others are reporting: `socket.once is not a function`.

When I use the `mongodb+srv://` connection protocol, I get these errors:
- `workerd/jsg/util.c++:274: error: e = kj/compat/tls.c++:256: failed: TLS peer's certificate is not trusted; reason = unable to get local issuer certificate`
- `ERR Error [EBADQUERY]QUERY _mongodb._tcp.MY_DATABASE_CLUSTER.mongodb.net`

I'm using:
- yarn@4.6.0 [with `nodeLinker: node-modules`]
- wrangler@^3.106
- mongodb@^6.13.0

**Reply by tobsn** [reply]
they're already working on that @DylanBulmer - [see 2 messages above yours](https://github.com/cloudflare/workerd/discussions/2721#discussioncomment-11996369)

**Reply by Imtiajrex** [reply]
> @JetLua @vicb I'll be working on adding TLSSocket.

any progress on the tlssocker?

**Reply by anonrig** [reply]
I've opened a PR but it's in progress. https://github.com/cloudflare/workerd/pull/3594

**Reply by anonrig** [reply]
TLSSocket and connect() has landed into main. It will be available to the general public in the next couple of weeks.

**Reply by alexbevi** [reply]
Very exciting @anonrig

**Reply by JetLua** [reply]
> TLSSocket and connect() has landed into main. It will be available to the general public in the next couple of weeks.

~~Latest Wrangler test result: `querySRV EBADQUERY _mongodb._tcp.yew.oqbm7gt.mongodb.net`~~

update: everything was perfect!

**Reply by sarge1989** [reply]
@JetLua Ran your minimal repro, mine's working on a deployed worker

![image](https://github.com/user-attachments/assets/3e9d4472-21eb-4e6b-a8c8-c651d59cb56b)

**Reply by JetLua** [reply]
Sorry, everything was perfect! Thank you.

**Reply by huw** [reply]
Working for me, including with Mongoose! Mongoose just needs a quick patch to add `package.json#exports`:

```diff
diff --git a/package.json b/package.json
index 46468fcb6a56ecd494c5a590b2940746a29c57ba..29a1a366f9e2adbdac15ee3dc3d43648d40eb3cd 100644
--- a/package.json
+++ b/package.json
@@ -112,6 +112,15 @@
   },
   "main": "./index.js",
   "types": "./types/index.d.ts",
+  "exports": {
+    ".": {
+      "workerd": "./index.js",
+      "node": "./index.js",
+      "browser": "./dist/browser.umd.js",
+      "types": "./types/index.d.ts",
+      "default": "./index.js"
+    }
+  },
   "engines": {
     "node": ">=16.20.1"
   },
```

(Otherwise, it’ll import the browser bundle. See Automattic/mongoose#14613)

**Reply by pawmanaloto** [reply]
> Working for me, including with Mongoose! Mongoose just needs a quick patch to add `package.json#exports`:
> 
> ```diff
> diff --git a/package.json b/package.json
> index 46468fcb6a56ecd494c5a590b2940746a29c57ba..29a1a366f9e2adbdac15ee3dc3d43648d40eb3cd 100644
> --- a/package.json
> +++ b/package.json
> @@ -112,6 +112,15 @@
>    },
>    "main": "./index.js",
>    "types": "./types/index.d.ts",
> +  "exports": {
> +    ".": {
> +      "workerd": "./index.js",
> +      "node": "./index.js",
> +      "browser": "./dist/browser.umd.js",
> +      "types": "./types/index.d.ts",
> +      "default": "./index.js"
> +    }
> +  },
>    "engines": {
>      "node": ">=16.20.1"
>    },
> ```
> 
> (Otherwise, it’ll import the browser bundle. See [Automattic/mongoose#14613](https://github.com/Automattic/mongoose/issues/14613))

How did you make it work? Do I need to do something or just wait for Mongoose to patch it?

[the remark this answers is not in the document]
**Reply by alexbevi** [reply]
Thanks for flagging @vicb. I've been following along closely and looking forward to testing this out now that the latest `workerd` is released with additional Node API support.

**Comment by ai-slave** [comment]
The moment i finally give up and switch to D1 from mongo you release THIS. But big thank you for finally doing this!!

**Reply by dustinsherwood** [reply]
Same

**Reply by ricky11** [reply]
how is your experienced  moving to d1 from mongo? d1 is sql i believe.  must be low latency for all over the world?

**Comment by equalogix** [comment]
I want to raise another related issue. First off there is a workaround see: [mongodb polyfill](https://github.com/jchoi2x/cf-mongodb-polyfills)
I have been using it in development since it was published and it mostly works. However, there is a bug that occurs when making certain requests, Mongodb throws: `BSONOffsetError`. I believe it is caused when the requests to the database returns a large object. In my case I get the error when making `$vectorSearch` request and getting back many documents. The exact error is:
```Mongodb vector search error: BSONOffsetError: Parsed documentSize (10568 bytes) does not match input length (4075 bytes). offset: 0```
I'm guessing that this error is not specific to the polyfill and will arise even when TLSSocket issue is resolved.

Another important point:
I am developing my app using ReactRouterv7, if I run the app locally using the `react-router dev` (basically using node-js), I don't get the errors. But if I run locally using `wrangler dev` or if I deploy the app, then I get the errors.

@alexbevi, seeing as you are from Mongodb, do you have any insight about this error, have you encountered it? The `BSONOffsetError` object itself is marked as `experimental` in the Mongodb Docs.

Finally I'm not the only one that is getting these errors, see [polyfill issue 4 ](https://github.com/jchoi2x/cf-mongodb-polyfills/issues/4)

**Reply by equalogix** [reply]
I was wrong, this bug has everything to do with the polyfill. But the good news is that I fixed it, and I am now able to use MongoDb Atlas from my Cloudflare worker in preview and production. I have made a pull-request in the polyfill's repo with changes, waiting for it to be reviewed and merged.

**Reply by alexbevi** [reply]
Thanks for staying on top of this @equalogix. I haven't had a chance to test the polyfills again (hopefully can do that soon), but FYI https://github.com/cloudflare/workerd/pull/3594 is currently being worked on as well (which should hopefully be the last piece of the puzzle to getting this working natively in workerd)

**Reply by ricky11** [reply]
Amazing work guys, this will be one less reason to leave MongoDB due to the deprecation of the DATA API, hoping for documentation around this as well, or a blog post to help us newbs.

**Comment by ManasMadrecha** [comment]
If everything's working fine in a production deployed worker, can we kindly have a blog or official documentation on it to so that googling "how to use mongodb with Cloudflare worker" shows that article and not GH discussion? 😊🙏🏻 Thanks for all your amazing work everyone.

**Reply by alexbevi** [reply]
Since I wrote up the initial post about why Workers/MongoDB didn't work together, I've gone ahead and written one that calls out that they DO work together now :)

https://alexbevi.com/blog/2025/03/25/cloudflare-workers-and-mongodb/

**Reply by gerhardcit** [reply]
@alexbevi , have you done this in a pages app with typescript?

This pops up in typecript?

![image](https://github.com/user-attachments/assets/21f91f3d-21d0-4488-9a4f-2f59f1f750fa)

**Reply by gerhardcit** [reply]
I get a lot of Error 1101, 
Worker threw exception

and then worker logs gives me this:
```
"exceptions": [
    {
      "name": "Error",
      "message": "The script will never generate a response.",
      "timestamp": 1743024718742
    }
  ],
  ```
  
  anyone from Cloudflare can enlighten me on what that means?
  its seems to happen when I send requests to fast. seems that when I want a few seconds all os good. But it's somehow random?
  
  ```typescript
  import { getMongoClient } from '$lib/mongo';
import type { PageServerLoad } from './$types';

export const load = (async () => {
    const startTime = Date.now();
    const v = '0.0.4';

    try {
        const client = await getMongoClient();

        const db = client.db('test');
        const coll = db.collection('test_workers');

        await coll.insertOne({ a: new Date() });

        const result = await coll.find({}).sort({ _id: -1 }).limit(1).next();
        
        const data = JSON.parse(JSON.stringify(result));
        return {
            ...data,
            timeMs: Date.now() - startTime,
            v
        }
    } catch (e: any) {
        console.error("error", e);
        return { _id: 'error', a: e.message, timeMs: Date.now() - startTime, v };
    }
}) satisfies PageServerLoad;
```

```typescript
import { env } from '$env/dynamic/private';
import { MongoClient, type MongoClientOptions } from 'mongodb';
let client = null;

export async function getMongoClient(): Promise<MongoClient> {
    const uri = env.MONGODB_URI;
 // surely I should not create a new client all the time?
    client ??= new MongoClient(uri, {
        maxPoolSize: 1,
        minPoolSize: 0,
        serverSelectionTimeoutMS: 5000,
    } as MongoClientOptions);
    await client.connect();
    return client;
};
````

**Reply by irvinebroque** [reply]
https://developers.cloudflare.com/workers/observability/errors/#the-script-will-never-generate-a-response-errors

**Reply by gerhardcit** [reply]
turns out that not reusing a client solves it

after some logging and tailing i discovered the error happens when 'reusing client' was used.
so now I create a new client on each request, or does workers handle this somehow under the hood?

it seems that client was not null on the 2nd request, but some how it lost the connection? (it's beyond my technical scope)
I would just like to know how this will effect the client connection count in Atlas.

```typescript
import { env } from '$env/dynamic/private';
import { MongoClient, type MongoClientOptions } from 'mongodb';
// let client: MongoClient | null = null;

export async function getMongoClient(): Promise<MongoClient> {
    // if (client) {
    //     console.log('reusing client');
    //     await client.connect();
    //     return client
    // }
    console.log("Attempting MongoDB connection");
    const uri = env.MONGODB_URI;

    const client = new MongoClient(uri, {
        maxPoolSize: 1,
        minPoolSize: 0,
        serverSelectionTimeoutMS: 5000,
    } as MongoClientOptions);
    await client.connect();
    console.log("Connected");
    return client;
};
```

**Reply by ricky11** [reply]
wont this create too many connections in atlas? each user may require 2-3 connections, and if your making a new non pooled connection every time, you will quickly exhaust the 500 connection cap.   I know workers are very short lived, but surely we can reuse the same connection.

**Reply by equalogix** [reply]
You need to explicitly close the connection when you are done with it with `client.close()`. You can (and should) keep a connection open and make multiple call to the db from it (as needed) but only for as long as the worker is processing its request.

**Reply by ricky11** [reply]
Yes was thinking the same.. I currently use lamba functions with mongdb, and functions last about 20-30 seconds, subsequent connections use the existing connection and provide super fast 10ms connectivity...  pooling is out of my technical expertise, but would be great if this supports cloudflare hyperdrive which keeps connections open (currently only works with sql  i believe)

**Reply by gerhardcit** [reply]
@ricky11 , been testing this with a Mongo Atlas (Flex) db.
added maxIdleTimeMS and using client.close() I cannot get the number of connections past 10 to show up on Atlas monitoring.

```typescript
const client = new MongoClient(uri, {
    maxPoolSize: 1,
    minPoolSize: 0,
    serverSelectionTimeoutMS: 5000,
    maxIdleTimeMS: 10000, // Close idle connections after 10 seconds
} as MongoClientOptions);
```
but I would love to hear some expertise around the architecture of mongo connections and how Atlas handles that.

**Reply by ricky11** [reply]
CF Pages returns a build error from one of my functions.. I have the flags set correctly in wrangler.toml within my functions directory.

Works beautifully locally with wrangler pages dev ./dist, but when pushed to cf, pages fails to build the project.

Any ideas?


14:32:38.016 | ✘ [ERROR] Could not resolve "mongodb"
-- | --
14:32:38.016 |  
14:32:38.016 | api/incrementdlPusher.js:1:38:
14:32:38.016 | 1 │ import { MongoClient, ObjectId } from 'mongodb';
14:32:38.016 | ╵                                       ~~~~~~~~~
14:32:38.016 |  
14:32:38.016 | You can mark the path "mongodb" as external to exclude it from the bundle, which will remove this error.

**Reply by ricky11** [reply]
Found that this is because pages uses ⛅️ wrangler 3.101.0, now i need to see how i can modify the version in pages.

**Comment by ricky11** [comment]
Has anyone gotten this to work with Pages Functions as oppose to native workers.. I think CF Pages Functions use a different version of wrangler which still fails.

**Reply by anonrig** [reply]
cc @vicb

**Reply by ricky11** [reply]
Found out that pages functions is pinned to 3.101, what we need is 3.106.  I was asking the guys on discord too.

![image](https://github.com/user-attachments/assets/1165330c-aa2d-493a-85b3-d993014f8379)

**Comment by huw** [comment]
Starting a new thread for issues with connection pooling. We’d like to re-use connections across invocations, because otherwise we add a ~1500ms cold start to every invocation. However, @gerhardcit noted that you can’t re-use connection pools (created by `mongodb.connect` or `mongodb.createConnection`) across invocations of the same worker, because it gives this error:

```
✘ [ERROR] A hanging Promise was canceled. This happens when the worker runtime is waiting for a Promise from JavaScript to resolve, but has detected that the Promise cannot possibly ever resolve because all code and events related to the Promise's I/O context have already finished.
✘ [ERROR] Uncaught (in response) Error: The script will never generate a response.
```

This isn’t that informative, because the actual issue is that a deeper exception gets caught, which causes the promises to not resolve. Firing up the debugger and breaking on caught exceptions gives us:

```
Exception has occurred: Error: Cannot perform I/O on behalf of a different request. I/O objects (such as streams, request/response bodies, and others) created in the context of one request handler cannot be accessed from a different request's handler. This is a limitation of Cloudflare Workers which allows us to improve overall performance. (I/O type: Writable)
  at Socket._writeGeneric (node-internal:internal_net:328:33)
    at Socket._write (node-internal:internal_net:355:10)
    at writeOrBuffer (node-internal:streams_writable:366:12)
    at _write (node-internal:streams_writable:308:10)
    at TLSSocket.write (node-internal:streams_writable:312:10)
    at _Connection.writeCommand (file:///<REDACTED>/.wrangler/tmp/dev-V18Y3f/index.js:28009:25)
    at async _Connection.sendWire (file:///<REDACTED>/.wrangler/tmp/dev-V18Y3f/index.js:27877:11)
    at async _Connection.sendCommand (file:///<REDACTED>/.wrangler/tmp/dev-V18Y3f/index.js:27920:22)
    at async _Connection.command (file:///<REDACTED>/.wrangler/tmp/dev-V18Y3f/index.js:27957:26)
    at async _Server2.command (file:///<REDACTED>/.wrangler/tmp/dev-V18Y3f/index.js:29563:25)
```

Okay, so it seems like a non-starter to actually re-use a connection in a Worker. Instead, let’s try use a singleton Durable Object to manage the connection pool. Durable Objects shut down about a minute after their last client disconnects, which should be enough to keep a connection pool around across multiple requests.

I added this to `wrangler.jsonc`:

```json
{
  "durable_objects": {
    "bindings": [
      {
        "name": "MONGO_DB_CONNECTION",
        "class_name": "MongoDBConnection"
      }
    ]
  },
  "migrations": [
    {
      "tag": "v1",
      "new_sqlite_classes": ["MongoDBConnection"]
    }
  ]
}
```

My `index.js` looks like this. I’ve provided a `locationHint` to my DO, based on where my MongoDB databases are:

```js
export { MongoDBConnection } from "./MongoDBConnection.js";

export default {
  async fetch(request, { MONGO_DB_CONNECTION, MONGO_DB_URI }) {
    // FWIW, you’ll probably want to do an auth check or similar here, to prevent wasting resources on fuzzing attacks
    // We can just forward the request straight onto the Durable Object
    const id = MONGO_DB_CONNECTION.idFromName(MONGO_DB_URI);
    const connection = MONGO_DB_CONNECTION.get(id, { locationHint: "apac" });
    return await connection.fetch(request);
  }
};
```

Then `MongoDBConnection.js` looks like this:

```js
import { DurableObject } from "cloudflare:workers";
import mongoose from "mongoose";

export class MongoDBConnection extends DurableObject {
  connection;

  constructor(ctx, env) {
    super(ctx, env);
    this.connection = mongoose.createConnection(env.MONGO_DB_URI, {
      // Decrease the selection timeout from 30s (default) to 5s, so that if something breaks, we don't wait around for it.
      serverSelectionTimeoutMS: 5000,
      // Time out idle connections after ~20 seconds, to avoid using up too much wall time (tweak this as needed for your use case)
      maxIdleTimeMS: 20 * 1000,
    });
  }

  async fetch(request) {
    // Wait for the connection to become available
    await this.connection.asPromise();
    // Do work here
    return new Response("OK");
  }
}
```

This code just transparently forwards all requests straight onto the DO, which can hold onto the connection pool across multiple requests and re-use it. My end-to-end timings went from ~2000ms to ~300ms after doing this!

**Reply by alexbevi** [reply]
DO Is the way to go for working with MongoDB from Workers. I wrote up some benchmarking results at https://alexbevi.com/blog/2025/04/11/performance-profiling-mongodb-on-cloudflare-workers/ with an example using the driver.

**Reply by gerhardcit** [reply]
Thanks @huw  and @alexbevi , this is "groundbreaking" in many ways.
It's the final cherry on the top of making mongo usefull in workers. (And it's been a blocker for many I think)
D1 as a db is great, but the mongo document flexibility outweighs that for us.

We started working our way back from pages to workers (all sveltekit projects) and one issue was developing with DO locally in the same dev space.
From what I see in workers the DO is integrated and no need for an separate DO. That also makes dev a lot easier.

We were toying with the idea of adding a MongoClient on the locals object (also sveltekit related I think.. never really done workers without sveltekit) But this opens up a better option.

Would like to hear feedback from other's on Yay or Nay..

**Reply by huw** [reply]
@alexbevi I’ve been messing around with different timeout parameters—I think the default `maxIdleTimeMS` (`0`) keeps the Durable Object open for longer than it needs to (~a minute after the most recent request, possibly longer). But I don’t have a deep enough understanding of MongoDB or DOs to figure out the optimal combination of parameters here.

**Reply by alexbevi** [reply]
@huw `maxIdleTimeMS` defaults to 0 and (from the [docs](https://www.mongodb.com/docs/drivers/node/current/fundamentals/connection/connection-options/)) _"Specifies the amount of time, in milliseconds, a connection can be idle before it's closed. Specifying 0 means no minimum."_.

> But I don’t have a deep enough understanding of MongoDB or DOs to figure out the optimal combination of parameters here.

I definitely don't have a deep enough understanding of DOs here, but the [connection options documentation for the Node driver](https://www.mongodb.com/docs/drivers/node/current/fundamentals/connection/connection-options/) lists the various tuneables you have access to.

**Reply by gerhardcit** [reply]
> Thanks @huw and @alexbevi , this is "groundbreaking" in many ways. It's the final cherry on the top of making mongo usefull in workers. (And it's been a blocker for many I think) D1 as a db is great, but the mongo document flexibility outweighs that for us.
> 
> We started working our way back from pages to workers (all sveltekit projects) and one issue was developing with DO locally in the same dev space. From what I see in workers the DO is integrated and no need for an separate DO. That also makes dev a lot easier.
> 
> We were toying with the idea of adding a MongoClient on the locals object (also sveltekit related I think.. never really done workers without sveltekit) But this opens up a better option.
> 
> Would like to hear feedback from other's on Yay or Nay..

Seems like I was mistaken in thinking Sveltekit and DurableObjects works together in development mode. See my comment here.
https://github.com/cloudflare/cloudflare-docs/issues/13062#issuecomment-1986928294

How I wouldn't like to get that working. `npm run dev` in a sveltekit project and be able to use a DurableObject. Surely that is not too much to ask?

**Reply by raininglemons** [reply]
@alexbevi which version of mongodb did you use for your blog post? Been testing something similar to your article, but get serialization issues (presumably with DO). Want to try and test as close to the working example you had :)

```
Could not serialize object of type "Db2". This type does not support serialization.
```

**Reply by alexbevi** [reply]
@raininglemons fairly certain it was [6.15.0](https://www.npmjs.com/package/mongodb/v/6.15.0)

**Reply by Imtiajrex** [reply]
> @alexbevi which version of mongodb did you use for your blog post? Been testing something similar to your article, but get serialization issues (presumably with DO). Want to try and test as close to the working example you had :)
> 
> ```
> Could not serialize object of type "Db2". This type does not support serialization.
> ```

You can't pass javascript Objects/function back and forth between DO and workers. only json serializable data can be passed

**Reply by raininglemons** [reply]
Ah @Imtiajrex that's it. I've misunderstood DO. I was trying to instantiate the client, then pass it to the calling worker. So really what I need to do is actually proxy all calls to the client through the DO, then return the response after it's resolved. Hero, thanks

**Reply by irvinebroque** [reply]
You can pass any structuredclone-able JS type between Workers and Durable Objects

https://developers.cloudflare.com/workers/runtime-apis/rpc/

**Reply by huw** [reply]
Let me know if any of you figure out how to deal with this:

<img width="299" alt="image" src="https://github.com/user-attachments/assets/fd929938-cfdb-4cb5-991e-5285b89de0ab" />

All of my requests return quite quickly, no issues, but because the Durable Object hangs around I get 2 issues with Workers Logs. First is, as above, the wall times for all of my invocations look much longer than I’d expect them to report. And second, the logs for the Durable Objects don’t show up in Workers Logs (either real-time or stored) until the Durable Object goes idle, which can take minutes. It makes debugging annoying. Evidently, the cause of this is the long timeout on the MongoDB connection, but I don’t want to shorten that.

**Reply by alberto-htl** [reply]
Still slower than Aws lambda right? It would be great to be able to reuse the connection natively between calls like lambda.

**Comment by flixyudh** [comment]
someone can tell me how to make a connection between cloudflare worker with mongoDB Atlas? i don't know how to get public IP of cloudflare worker to be added into mongoDB Atlas `network access` dashboard.

**Reply by ricky11** [reply]
I dont think CF will give you a fixed IP, unless I missed that part, but you will have to open to 0.0.0.0 which will allow all access.

**Reply by thomasnet-mc** [reply]
Cloudflare provides their IP ranges here: https://www.cloudflare.com/en-gb/ips/

**Reply by jkauszler** [reply]
Has anyone had luck connecting remotely with these public addresses? I've only been able to get 0.0.0.0 to work.

**Reply by tobsn** [reply]
> Has anyone had luck connecting remotely with these public addresses? I've only been able to get 0.0.0.0 to work.

you can't basically, I asked Cloudflare directly. if you use workers the remote IP addresses are all over the place from ALL their IP addresses they have, all ranges — the IP list is something separate and does not apply to workers. (not what Im saying, quoting their support)
You are forced to open it to any IP range. I switched to Cloudflares D1 and Supabase... not worth the pain in the butt trying to make this work. I guess it's in benefit of Cloudflare.

**Reply by jkauszler** [reply]
> if you use workers the remote IP addresses are all over the place from ALL their IP addresses they have, all ranges

This makes sense if they're leveraging each of their nodes around the globe for compute (workers). It's definitely not ideal though for an access security perspective.  Thanks for the report, @tobsn 

If anyone has found ways to implement a network allow list on mongo that doesn't also redirect from a single proxy, I'd love to hear how. Curious to see if there's a clever solution, either with existing CF services (R2, KV, DO, etc) or some external service. 

@alexbevi you've been exploring cool ways integrate CF and Mongo together. Any ideas here?

**Reply by tobsn** [reply]
fyi I think the whole mongodb and Cloudflare implementation issue is a very very niche problem as this wasnt possible in the beginning so most have already skipped it long ago. if you ask claude/etc. it will actually tell you its not possible as this was status quo for so long. you could use D1 and then http into D1 and sync it into mongodb?

**Reply by anonrig** [reply]
cc @thomasgauvin

**Comment by ricky11** [comment]
Tangent:  been using the Postmark node sdk with workers for a couple of weeks, but today the worker returned:

"There is no suitable adapter to dispatch the request since :\n- adapter xhr is not supported by the environment\n- adapter http is not available in the build"

Postmark does have a API, but  cumbersome, any chance of looking into Postmark support?

**Comment by 66Bunz** [comment]
I tried to set up an integration between MongoDB and Cloudflare Workers (free plan), following Alex Bevilacqua's tutorials, this discussion and other threads around the internet, tried to mess around with compatibility dates and flags, but I still get this following error only when deploying, it's working fine with `wrangler dev` on my machine:

```
MongoDB connection error: MongoServerSelectionError: [unenv] net.createConnection is not implemented yet!
```

am I missing something? I can see that the `createConnection` is not available yet (https://developers.cloudflare.com/workers/runtime-apis/nodejs/http/#request) but I guess that for someone it works, I just cannot figure out what the missing part is. 

If someone could help me I'll be grateful.

**Reply by anonrig** [reply]
please enable nodejs compat flag, and set a recent compat date.

**Reply by alexbevi** [reply]
@66Bunz If you're going off the article I have at https://alexbevi.com/blog/2025/04/11/performance-profiling-mongodb-on-cloudflare-workers/#default-workers-experience I've got a `wrangler.jsonc` and test script that should allow you to verify this.

**Reply by 66Bunz** [reply]
> please enable nodejs compat flag, and set a recent compat date.


This is my `wrangler.toml`

```toml
name = "droptop-api"
main = "src/index.js"
workers_dev = false
compatibility_date = "2025-09-15"
compatibility_flags = ["nodejs_compat"]

[observability]
enabled = true

[[durable_objects.bindings]]
name = "MONGODB_DURABLE_OBJECT"
class_name = "MongoDBDurableConnector"

[[migrations]]
tag = "v1"
new_sqlite_classes = ["MongoDBDurableConnector"]
```

**Reply by 66Bunz** [reply]
> @66Bunz If you're going off the article I have at https://alexbevi.com/blog/2025/04/11/performance-profiling-mongodb-on-cloudflare-workers/#default-workers-experience I've got a `wrangler.jsonc` and test script that should allow you to verify this.

Creating a new worker with your example works. I'll try to create a new one for mine too, maybe this solves the issues.

**Reply by 66Bunz** [reply]
Idk why when I stopped using the `wrangler.toml` and started with `wrangler.jsonc` file everything works.

**Comment by tobsn** [comment]
@66Bunz 

```
import { MongoClient, MongoClientOptions, UpdateResult } from 'mongodb';
client = new MongoClient(mongoUri, {
        maxPoolSize: 5,
        minPoolSize: 0,
        serverSelectionTimeoutMS: 10000,
        connectTimeoutMS: 10000,
        socketTimeoutMS: 10000,
        maxIdleTimeMS: 30000,
        minHeartbeatFrequencyMS: 2000,
      } as MongoClientOptions);
```

wranger.toml
```
compatibility_flags = ["nodejs_compat_v2"]
compatibility_date = "2024-03-20"
```

packages.json
```
  "dependencies": {
    "mongodb": "^6.18.0",
```

Quick copy paste from one of a quick worker I did last week, its vibe coded because its just a 1:1 proxy of data no handling and I did not want to waste time on it - it used the example code from the blog that shows how to implement the socket changes.
I had not seen the compat date until now because it worked fine...  hope it helps a bit.

**Reply by 66Bunz** [reply]
Idk why when I stopped using the wrangler.toml and started with wrangler.jsonc file everything works. Even though you have been using a toml file, idk why mine wasn't working

**Reply by tobsn** [reply]
im not sure either what exactly works... I tried for a while earlier this year to get it to work and I just couldn't. no example, no config combination worked. couple weeks ago I fed all the data into Claude Code and had it figure it out and it ran like 10 different combinations and that somehow then started to work.

**Comment by mikenomitch** [comment]
Resolving as this is now doable with nodejs_compat

**Reply by ricky11** [reply]
Yes it works work Cloudflare Workers, but does not work in Cloudflare pages functions -- reason is the workerd that is used in cloudflare functions is pinned to use a older version hence even with the nodejs_compact this will not work in functions.

**Reply by MasFana** [reply]
Does it work on Framework Template like Solid,Svelte ?

**Comment by ricky11** [comment]
Hi, What are you going doing to reuse your connection pool in workers?  
Every worker spawns it own connection, which will blow past the 500 connection limit.

Tks.

**Reply by tobsn** [reply]
DO

**Reply by alexbevi** [reply]
https://alexbevi.com/blog/2025/04/11/performance-profiling-mongodb-on-cloudflare-workers/#maintaining-state-with-durable-objects

**Reply by ricky11** [reply]
awesome, does hyperdrive support mongodb or still postgress. will try DO methods.

**Reply by alexbevi** [reply]
Hyperdrive is just Postgres and Mysql I believe
